File: solutions/typescript/bank-account/1/bank-account.ts
//
// This is only a SKELETON file for the 'Bank Account' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class ValueError extends Error {
  constructor() {
    super('Bank account error')
  }
}

export class BankAccount {
  currentBalance: number

  constructor() {
    this.currentBalance = NaN;
  }

  open(): void {
    if (!isNaN(this.currentBalance)) {
      throw new ValueError();
    }
    this.currentBalance = 0;
  }

  close(): void {
    if (isNaN(this.currentBalance)) {
      throw new ValueError();
    }
    this.currentBalance = NaN;
  }

  deposit(amount: number): void {
    if (isNaN(this.currentBalance) || amount < 0) {
      throw new ValueError();
    }
    this.currentBalance += amount;
  }

  withdraw(amount: number): void{
    if (isNaN(this.currentBalance) || amount < 0 || amount > this.currentBalance) {
      throw new ValueError();
    }
    this.currentBalance -= amount;
  }

  get balance(): unknown {
    if (isNaN(this.currentBalance)) {
      throw new ValueError();
    }
    return this.currentBalance;
  }
}

File: solutions/go/bank-account/1/bank_account.go
package account

import (
	"sync"
)

// Define the Account type here.
type Account struct {
	balance int64
	open    bool
	mu      *sync.Mutex
}

func Open(amount int64) *Account {
	if amount < 0 {
		return nil
	}
	return &Account{balance: amount, open: true, mu: &sync.Mutex{}}
}

func (a *Account) Balance() (int64, bool) {
	a.mu.Lock()
	defer a.mu.Unlock()
	if !a.open {
		return 0, false
	}
	return a.balance, true
}

func (a *Account) Deposit(amount int64) (int64, bool) {
	a.mu.Lock()
	defer a.mu.Unlock()
	if !a.open || a.balance+amount < 0 {
		return 0, false
	}
	a.balance += amount
	return a.balance, true
}

func (a *Account) Close() (int64, bool) {
	a.mu.Lock()
	defer a.mu.Unlock()
	if !a.open {
		return 0, false
	}
	balance := a.balance
	a.balance = 0
	a.open = false
	return balance, true
}
